dbos-rules.ts

```diff
@@ -0,0 +1,442 @@
+import { TypeChecker } from "typescript";
+import * as tslintPlugin from "@typescript-eslint/eslint-plugin";
+import { ESLintUtils, TSESLint, TSESTree, ParserServicesWithTypeInformation } from "@typescript-eslint/utils";
+
+import {
+  createWrappedNode, Node, FunctionDeclaration,
+  CallExpression, ConstructorDeclaration, ClassDeclaration,
+  MethodDeclaration
+} from "ts-morph";
+
+// Should I find TypeScript variants of these?
+const secPlugin = require("eslint-plugin-security");
+const noSecrets = require("eslint-plugin-no-secrets");
+
+type EslintNode = TSESTree.Node;
+type EslintContext = TSESLint.RuleContext<string, unknown[]>;
+
+//////////////////////////////////////////////////////////////////////////////////////////////////// Here is my `ts-morph` linting code:
+
+////////// These are some shared types and values used throughout the code
+
+// TODO: support `FunctionExpression` and `ArrowFunction` too
+type FunctionOrMethod = FunctionDeclaration | MethodDeclaration | ConstructorDeclaration;
+
+// This returns `undefined` if there is no error message to emit; otherwise, it returns a key to the `ERROR_MESSAGES` map
+type DetChecker = (node: Node, fn: FunctionOrMethod, isLocal: (name: string) => boolean) => string | undefined;
+
+type GlobalTools = {eslintContext: EslintContext, parserServices: ParserServicesWithTypeInformation, typeChecker: TypeChecker};
+let GLOBAL_TOOLS: GlobalTools | undefined = undefined;
+
+// These included `Transaction` and `TransactionContext` respectively before!
+const DETERMINISTIC_DECORATORS = new Set(["Workflow"]);
+const TYPES_YOU_CAN_AWAIT_UPON_IN_DETERERMINISTIC_FUNCTIONS = new Set(["WorkflowContext"]);
+const ERROR_MESSAGES = makeErrorMessageSet();
+
+////////// This is the set of error messages that can be emitted
+
+function makeErrorMessageSet(): Map<string, string> {
+  const makeDateMessage = (bannedCall: string) => `Calling ${bannedCall} is banned \
+(consider using \`@dbos-inc/communicator-datetime\` for consistency and testability)`;
+
+  const bcryptMessage = "Avoid using `bcrypt`, which contains native code. Instead, use `bcryptjs`. \
+Also, some `bcrypt` functions generate random data and should only be called from communicators";
+
+  const validTypeSetString = [...TYPES_YOU_CAN_AWAIT_UPON_IN_DETERERMINISTIC_FUNCTIONS].map((name) => `\`${name}\``).join(", ");
+
+  // The keys are the ids, and the values are the messages themselves
+  return new Map([
+    ["globalModification", "This is a global modification relative to the workflow declaration"],
```

Comment:
Warning: DBOS operations may not mutate global variables.

Reply:
Done